File: SosnowskyHogweed/SosnowskyHogweed.cpp
#include "SosnowskyHogweed.h"

SosnowskyHogweed::SosnowskyHogweed(World *world, int x, int y)
    : Plant(SOSNOWSKY_HOGWEED_STRENGTH,
            SOSNOWSKY_HOGWEED_SYMBOL,
            SOSNOWSKY_HOGWEED_COLOR,
            world,
            x,
            y) {}

void SosnowskyHogweed::killNeighbour(int x, int y) {
  if (!this->getWorld()->isPositionCorrect(x, y)) return;

  Organism* organism = this->getWorld()->getOrganismAtPosition(this, x, y);
  if (organism == nullptr || organism->getInitiative() == 0) return;

  organism->setAlive(false);
  this->getWorld()->addLog("%s POS:(%d, %d) killed %s POS:(%d, %d)",
                           this->getType().c_str(),
                           this->getPosition().getX(),
                           this->getPosition().getY(),
                           organism->getType().c_str(),
                           organism->getPosition().getX(),
                           organism->getPosition().getY());
}

void SosnowskyHogweed::action() {
  int x = this->getPosition().getX();
  int y = this->getPosition().getY();

  this->killNeighbour(x + 1, y);
  this->killNeighbour(x - 1, y);
  this->killNeighbour(x, y + 1);
  this->killNeighbour(x, y - 1);

  Plant::action();
}

void SosnowskyHogweed::collision(Organism* organism) {
  organism->setAlive(false);
  this->getWorld()->addLog("%s POS:(%d, %d) killed %s POS:(%d, %d)",
                           this->getType().c_str(),
                           this->getPosition().getX(),
                           this->getPosition().getY(),
                           organism->getType().c_str(),
                           organism->getPosition().getX(),
                           organism->getPosition().getY());
}

std::string SosnowskyHogweed::getType() const { return "SosnowskyHogweed"; }

Organism* SosnowskyHogweed::createChild(int x, int y, int parent1Id, int parent2Id) {
  return new SosnowskyHogweed(this->getWorld(), x, y);
}

File: src/SosnowskyHogweed/SosnowskyHogweed.java
package SosnowskyHogweed;

import Field.Field;
import NightshadeBerries.NightshadeBerries;
import Organism.Organism;
import Organism.OrganismData;
import Plant.Plant;
import Settings.Settings;
import World.World;

public class SosnowskyHogweed extends Plant {

    public SosnowskyHogweed(World world, Field field) {
        super(new OrganismData(world.getNextOrganismId(), Settings.SOSNOWSKY_HOGWEED_INITIATIVE, Settings.SOSNOWSKY_HOGWEED_STRENGTH, Settings.SOSNOWSKY_HOGWEED_IMAGE, -1, -1), world, field);
    }

    @Override
    public void action() {
        for (Field field : this.currentField.getNeighbours()) {
            if (field.getOrganism() != null && field.getOrganism().isAnimal() && field.getOrganism().getData().isAlive() && field.getOrganism().getData().isBorn()) {
                field.getOrganism().getData().setAlive(false);
                this.world.addLog(this.getType() + " killed " + field.getOrganism().getType());
            }
        }

        super.action();
    }

    @Override
    public void collision(Organism attacker) {
        if (attacker.isAnimal()) {
            attacker.getData().setAlive(false);
        }
    }

    @Override
    public String getType() {
        return "SosnowskyHogweed";
    }

    @Override
    public void reproduce(Field field) {
        Organism sosnowskyHogweed = new SosnowskyHogweed(this.world, field);
        this.world.addOrganism(sosnowskyHogweed);
        this.world.addLog(this.getType() + " reproduced (" + field.getNumber() + ")");
    }

}
